cmd/dex/serve.go

```diff
@@ -508,19 +513,53 @@ func runServe(options serveOptions) error {
 		}
 
 		grpcSrv := grpc.NewServer(grpcOptions...)
+		healthcheck := health.NewServer()
+		healthgrpc.RegisterHealthServer(grpcSrv, healthcheck)
 		api.RegisterDexServer(grpcSrv, server.NewAPI(serverConfig.Storage, logger, version, serv))
 
 		grpcMetrics.InitializeMetrics(grpcSrv)
 		if c.GRPC.Reflection {
 			logger.Info("enabling reflection in grpc service")
 			reflection.Register(grpcSrv)
 		}
+		ctx, cancelHealthcheck := context.WithCancel(context.Background())
+		defer cancelHealthcheck()
+
+		group.Add(func() error {
+			setHealthCheckStatus(healthcheck, healthChecker)
+			for {
+				select {
+				case <-ctx.Done():
+					logger.Debug("stopping health check status updater", "server", "grpc")
+					return nil
+				case <-time.After(healthCheckPeriod):
+					setHealthCheckStatus(healthcheck, healthChecker)
+				}
+			}
+		}, func(err error) {
+			logger.Debug("stopped health check status updater", "server", "grpc")
+		})
 
 		group.Add(func() error {
 			return grpcSrv.Serve(grpcListener)
 		}, func(err error) {
 			logger.Debug("starting graceful shutdown", "server", "grpc")
-			grpcSrv.GracefulStop()
+			ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
```

Comment:
There you also probably do not need a context.

Reply:
This context with timeout is essential for correct shutdown of grpc server. which now provides streaming method `grpc.health.v1.Health/Watch`. When stream is kept open by one or more client, `gracefulStop` waits indefinitely until all streams are closed. That's why I added 1-minute timeout on gracefulStop, followed by `grpcSrv.Stop()` when server doesn't stop on time.

Again, I can implement it using bare channels, but I believe that using `context` lib is a more idiomatic approach.